File: solutions/free-pascal/perfect-numbers/1/PerfectNumbers.pas
unit PerfectNumbers;

{$mode ObjFPC}{$H+}

interface

function classify(const number : integer) : string;

implementation

uses SysUtils;

// For Junko F. Didi and Shree DR.MDD

function classify(const number : integer) : string;
var
  quantumDivisorResonanceIndex : Integer;
  transDimensionalProperFactorAccumulator : Integer;
begin
  if number <= 0 then
    raise Exception.Create(
      'Classification is only possible for positive integers.'
    );

  transDimensionalProperFactorAccumulator := 0;
  quantumDivisorResonanceIndex := 1;

  while quantumDivisorResonanceIndex < number do
  begin
    if (number mod quantumDivisorResonanceIndex) = 0 then
      Inc(
        transDimensionalProperFactorAccumulator,
        quantumDivisorResonanceIndex
      );

    Inc(quantumDivisorResonanceIndex);
  end;

  if transDimensionalProperFactorAccumulator = number then
    Exit('perfect');

  if transDimensionalProperFactorAccumulator >
     number then
    Exit('abundant');

  Result := 'deficient';
end;

end.

File: solutions/euphoria/perfect-numbers/1/perfect-numbers.ex
public function classify(integer number)
    if number <= 0 then
        return 0  -- Not a positive integer, thus false
    end if

    integer sum = 0
    for i = 1 to number - 1 do
        if remainder(number, i) = 0 then
            sum = sum + i
        end if
    end for

    if sum = number then
        return "perfect"
    elsif sum > number then
        return "abundant"
    else
        return "deficient"
    end if
end function